HDDS-2240. Command line tool for OM HA.

hanishakoneru opened the pull request:
A command line tool (ozone omha) to get information related to OM HA. 
This Jira proposes to add the getServiceState option for OM HA which lists all the OMs in the service and their corresponding Ratis server roles (LEADER/ FOLLOWER). 
We can later add more options to this tool.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 427 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| 0 | shelldocs | 0 | Shelldocs was not available. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -1 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 26 | Maven dependency ordering for branch |
| -1 | mvninstall | 39 | hadoop-hdds in trunk failed. |
| -1 | mvninstall | 36 | hadoop-ozone in trunk failed. |
| -1 | compile | 21 | hadoop-hdds in trunk failed. |
| -1 | compile | 14 | hadoop-ozone in trunk failed. |
| +1 | checkstyle | 59 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 897 | branch has no errors when building and testing our client artifacts. |
| -1 | javadoc | 22 | hadoop-hdds in trunk failed. |
| -1 | javadoc | 19 | hadoop-ozone in trunk failed. |
| 0 | spotbugs | 996 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| -1 | findbugs | 34 | hadoop-hdds in trunk failed. |
| -1 | findbugs | 19 | hadoop-ozone in trunk failed. |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 17 | Maven dependency ordering for patch |
| -1 | mvninstall | 35 | hadoop-hdds in the patch failed. |
| -1 | mvninstall | 37 | hadoop-ozone in the patch failed. |
| -1 | compile | 25 | hadoop-hdds in the patch failed. |
| -1 | compile | 18 | hadoop-ozone in the patch failed. |
| -1 | cc | 25 | hadoop-hdds in the patch failed. |
| -1 | cc | 18 | hadoop-ozone in the patch failed. |
| -1 | javac | 25 | hadoop-hdds in the patch failed. |
| -1 | javac | 18 | hadoop-ozone in the patch failed. |
| -0 | checkstyle | 29 | hadoop-ozone: The patch generated 2 new + 0 unchanged - 0 fixed = 2 total (was 0) |
| +1 | mvnsite | 0 | the patch passed |
| +1 | shellcheck | 32 | There were no new shellcheck issues. |
| +1 | whitespace | 1 | The patch has no whitespace issues. |
| +1 | shadedclient | 824 | patch has no errors when building and testing our client artifacts. |
| -1 | javadoc | 21 | hadoop-hdds in the patch failed. |
| -1 | javadoc | 18 | hadoop-ozone in the patch failed. |
| -1 | findbugs | 32 | hadoop-hdds in the patch failed. |
| -1 | findbugs | 19 | hadoop-ozone in the patch failed. |
||| _ Other Tests _ |
| -1 | unit | 27 | hadoop-hdds in the patch failed. |
| -1 | unit | 25 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 33 | The patch does not generate ASF License warnings. |
| | | 3016 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.0 Server=19.03.0 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1586 |
| Optional Tests | dupname asflicense mvnsite unit shellcheck shelldocs compile javac javadoc mvninstall shadedclient findbugs checkstyle cc |
| uname | Linux ca7bea350dc9 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / a3fe404 |
| Default Java | 1.8.0_222 |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/1/artifact/out/branch-mvninstall-hadoop-hdds.txt |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/1/artifact/out/branch-mvninstall-hadoop-ozone.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/1/artifact/out/branch-compile-hadoop-hdds.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/1/artifact/out/branch-compile-hadoop-ozone.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/1/artifact/out/branch-javadoc-hadoop-hdds.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/1/artifact/out/branch-javadoc-hadoop-ozone.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/1/artifact/out/branch-findbugs-hadoop-hdds.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/1/artifact/out/branch-findbugs-hadoop-ozone.txt |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/1/artifact/out/patch-mvninstall-hadoop-hdds.txt |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/1/artifact/out/patch-mvninstall-hadoop-ozone.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/1/artifact/out/patch-compile-hadoop-hdds.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/1/artifact/out/patch-compile-hadoop-ozone.txt |
| cc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/1/artifact/out/patch-compile-hadoop-hdds.txt |
| cc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/1/artifact/out/patch-compile-hadoop-ozone.txt |
| javac | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/1/artifact/out/patch-compile-hadoop-hdds.txt |
| javac | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/1/artifact/out/patch-compile-hadoop-ozone.txt |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/1/artifact/out/diff-checkstyle-hadoop-ozone.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/1/artifact/out/patch-javadoc-hadoop-hdds.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/1/artifact/out/patch-javadoc-hadoop-ozone.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/1/artifact/out/patch-findbugs-hadoop-hdds.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/1/artifact/out/patch-findbugs-hadoop-ozone.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/1/artifact/out/patch-unit-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/1/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/1/testReport/ |
| Max. process+thread count | 305 (vs. ulimit of 5500) |
| modules | C: hadoop-ozone/common hadoop-ozone/ozone-manager U: hadoop-ozone |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/1/console |
| versions | git=2.7.4 maven=3.3.9 shellcheck=0.4.6 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

dineshchitlangia commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManagerHA.java:
```diff
@@ -0,0 +1,76 @@
+    Map<String, String> serviceStates = getServiceStates(conf, serviceId);
+    for (String nodeId : serviceStates.keySet()) {
+      System.out.println(nodeId + " : " + serviceStates.get(nodeId));
+    }
```
It will be better to use entrySet() instead of keySet() for performance. Although this method will not be doing extensive amount of work, I believe it is still worth making this change.

hanishakoneru replied:
Done.

dineshchitlangia commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManagerHA.java:
```diff
@@ -0,0 +1,76 @@
+    hidden = true, description = "Command line tool for OM HA.",
+    versionProvider = HddsVersionProvider.class,
+    mixinStandardHelpOptions = true)
+public class OzoneManagerHA extends GenericCli {
```
Declare this class as final.

dineshchitlangia commented on hadoop-ozone/common/src/main/bin/ozone:
```diff
@@ -55,6 +55,7 @@ function hadoop_usage
   hadoop_add_subcommand "version" client "print the version"
   hadoop_add_subcommand "dtutil" client "operations related to delegation tokens"
   hadoop_add_subcommand "upgrade" client "HDFS to Ozone in-place upgrade tool"
+  hadoop_add_subcommand "omha" client "OM HA tool"
```
NIT: Have we named this command `omha` because we have plans in future to also add similar command for SCM?
If not, how about naming the command as `haadmin` , `omadmin`, `admin`.
Not a deal breaker but was just thinking if we can keep it similar to hdfs.

hanishakoneru replied:
Changed the command to "ozone admin om". 
SCM HA is also in the pipeline sometime in the future. 
As Marton also suggested, we can have all admin commands under this. This will include SCMCli also.

anuengineer commented on hadoop-ozone/common/src/main/proto/OzoneManagerProtocol.proto:
```diff
@@ -1097,11 +1097,34 @@ message UpdateGetS3SecretRequest {
+    // A client-to-OM RPC to get ratis server states of OMs
+    rpc getServiceState(OMServiceId)
+          returns(ServiceState);
 }
```
What is the use case for this RPC? it is used only from the OM HA tool? or is some thing that clients need to know when communicating with the OM? if clients need to know this info, then clients already make a call called getServiceList and this info is more appropriate to be part of that call? Then both clients and tools like OM HA can get to that information pretty easily.

hanishakoneru replied:
Clients do not need this information. 
I have added a new OMRequest to get the OM server roles. Did not add it to getServiceList api because getServiceList is called when creating RpcClient (so all clients make this call when starting up). But adding the server roles also to this api would mean one extra Ratis call to get the server roles which is not required by all clients. Hence, kept the two calls separate.

anuengineer replied:
Sorry did not see this comment. How can a client communicate to OM? Does it not need to send the request to the leader ?

anuengineer commented on hadoop-ozone/common/src/main/proto/OzoneManagerProtocol.proto:
```diff
@@ -1097,11 +1097,34 @@ message UpdateGetS3SecretRequest {
     rpc submitRequest(OMRequest)
           returns(OMResponse);
+
+    // A client-to-OM RPC to get ratis server states of OMs
```
Also just wondering, is there a reason to add a new RPC and not use the submitRequet pattern -- that is add the message to the OMRequest / OMResponse pattern?

hanishakoneru replied:
You are right. There is no need for new RPC. I thought we could keep the admin protocols separate from the client protocols. But I think we can just use the OMRequest for now. Updated it.

elek commented:
Thank you very much to open this pull request.

During the weekend the Ozone source code has been moved out from [apache/hadoop](https://github.com/apache/hadoop) repository to [apache/hadoop-ozone](https://github.com/apache/hadoop-ozone) repository.

This git commits are rewritten, but the branch of this pull request is also transformed (state of Saturday morning), you can use the new, migrated branch to recreate this pull request.

Your pull request is important for us: **Can you please re-create your pull request in the new repository?**

**1**. Create a new fork of https://github.com/apache/hadoop-ozone

**2**. Clone it and have both your fork and the apache repo as remotes:

```
git clone git@github.com:hanishakoneru/hadoop-ozone.git
cd hadoop-ozone
git remote add apache git@github.com:apache/hadoop-ozone.git
git fetch apache
```

**3**. Fetch your migrated branch and push it to your fork.

```
git checkout -b HDDS-2240 apache/HDDS-2240
git push origin HDDS-2240
```

**4**. And create the new pull request on the new repository.

https://github.com/apache/hadoop-ozone/compare/master...hanishakoneru:HDDS-2240?expand=1

If you need more information, please check [this](https://cwiki.apache.org/confluence/display/HADOOP/Hadoop+Ozone+source+tree+split) wiki page or contact with me (my github user name + apache.org).

Thank you, and sorry for the inconvenience.

hanishakoneru commented:
@elek Thank you for your comments and suggestions.
1. I have changed the command to "ozone admin om". We can add scmcli also to the same _ozone admin_ command. Also followed the cli conventions. The command would look like 
_ozone admin om getserviceroles --service-id_
2. Updated the protocol and using the a new OMRequest to get the service roles.

hanishakoneru commented:
Thank you @dineshchitlangia @elek and @anuengineer for the review and comments.
I have migrated the Pull Request to the new hadoop-ozone repo - https://github.com/apache/hadoop-ozone/pull/9

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 42 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| 0 | shelldocs | 1 | Shelldocs was not available. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -1 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 35 | Maven dependency ordering for branch |
| -1 | mvninstall | 35 | hadoop-hdds in trunk failed. |
| -1 | mvninstall | 38 | hadoop-ozone in trunk failed. |
| -1 | compile | 24 | hadoop-hdds in trunk failed. |
| -1 | compile | 19 | hadoop-ozone in trunk failed. |
| +1 | checkstyle | 62 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 813 | branch has no errors when building and testing our client artifacts. |
| -1 | javadoc | 27 | hadoop-hdds in trunk failed. |
| -1 | javadoc | 24 | hadoop-ozone in trunk failed. |
| 0 | spotbugs | 932 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| -1 | findbugs | 38 | hadoop-hdds in trunk failed. |
| -1 | findbugs | 25 | hadoop-ozone in trunk failed. |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 23 | Maven dependency ordering for patch |
| -1 | mvninstall | 40 | hadoop-hdds in the patch failed. |
| -1 | mvninstall | 41 | hadoop-ozone in the patch failed. |
| -1 | compile | 29 | hadoop-hdds in the patch failed. |
| -1 | compile | 23 | hadoop-ozone in the patch failed. |
| -1 | cc | 29 | hadoop-hdds in the patch failed. |
| -1 | cc | 23 | hadoop-ozone in the patch failed. |
| -1 | javac | 29 | hadoop-hdds in the patch failed. |
| -1 | javac | 23 | hadoop-ozone in the patch failed. |
| -0 | checkstyle | 34 | hadoop-ozone: The patch generated 27 new + 0 unchanged - 0 fixed = 27 total (was 0) |
| +1 | mvnsite | 0 | the patch passed |
| +1 | shellcheck | 32 | There were no new shellcheck issues. |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 737 | patch has no errors when building and testing our client artifacts. |
| -1 | javadoc | 27 | hadoop-hdds in the patch failed. |
| -1 | javadoc | 24 | hadoop-ozone in the patch failed. |
| -1 | findbugs | 36 | hadoop-hdds in the patch failed. |
| -1 | findbugs | 26 | hadoop-ozone in the patch failed. |
||| _ Other Tests _ |
| -1 | unit | 33 | hadoop-hdds in the patch failed. |
| -1 | unit | 30 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 36 | The patch does not generate ASF License warnings. |
| | | 2577 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.3 Server=19.03.3 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1586 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle shellcheck shelldocs cc |
| uname | Linux 220484ce9b24 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 5f4641a |
| Default Java | 1.8.0_222 |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/2/artifact/out/branch-mvninstall-hadoop-hdds.txt |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/2/artifact/out/branch-mvninstall-hadoop-ozone.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/2/artifact/out/branch-compile-hadoop-hdds.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/2/artifact/out/branch-compile-hadoop-ozone.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/2/artifact/out/branch-javadoc-hadoop-hdds.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/2/artifact/out/branch-javadoc-hadoop-ozone.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/2/artifact/out/branch-findbugs-hadoop-hdds.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/2/artifact/out/branch-findbugs-hadoop-ozone.txt |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/2/artifact/out/patch-mvninstall-hadoop-hdds.txt |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/2/artifact/out/patch-mvninstall-hadoop-ozone.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/2/artifact/out/patch-compile-hadoop-hdds.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/2/artifact/out/patch-compile-hadoop-ozone.txt |
| cc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/2/artifact/out/patch-compile-hadoop-hdds.txt |
| cc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/2/artifact/out/patch-compile-hadoop-ozone.txt |
| javac | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/2/artifact/out/patch-compile-hadoop-hdds.txt |
| javac | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/2/artifact/out/patch-compile-hadoop-ozone.txt |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/2/artifact/out/diff-checkstyle-hadoop-ozone.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/2/artifact/out/patch-javadoc-hadoop-hdds.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/2/artifact/out/patch-javadoc-hadoop-ozone.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/2/artifact/out/patch-findbugs-hadoop-hdds.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/2/artifact/out/patch-findbugs-hadoop-ozone.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/2/artifact/out/patch-unit-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/2/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/2/testReport/ |
| Max. process+thread count | 451 (vs. ulimit of 5500) |
| modules | C: hadoop-ozone/common hadoop-ozone/client hadoop-ozone/ozone-manager hadoop-ozone/tools U: hadoop-ozone |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1586/2/console |
| versions | git=2.7.4 maven=3.3.9 shellcheck=0.4.6 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

elek commented:
Thanks the migration @hanishakoneru (and BTW, thanks the update, I will check it). I am closing this one as we have the new one.
